articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,241 @@
+- LIVELINESS_MANUAL_TOPIC - The signal that establishes a Topic is alive is at the Topic level. Only publishing a message on the Topic or an explicit signal from the application to assert liveliness on the Topic will mark the Topic as being alive.
+
+Services will support the following levels of liveliness.
+- LIVELINESS_DEFAULT - Use the ROS specified default for liveliness (which is LIVELINESS_AUTOMATIC).
+- LIVELINESS_AUTOMATIC - The signal that establishes a Service Owner is alive comes from the ROS rmw layer.
+- LIVELINESS_MANUAL_NODE - The signal that establishes a Service is alive is at the node level. A message on any outgoing channel on the node or an explicit signal from the application to assert liveliness on the node will mark all outgoing channels on the node as being alive.
+- LIVELINESS_MANUAL_SERVICE - The signal that establishes a Service is alive is at the Service level. Only sending a response on the Service or an explicit signal from the application to assert liveliness on the Service will mark the Service as being alive.
+
+In order for a Subscriber to listen to a Publisher's Topic the level of liveliness tracking they request must be equal or less verbose than the level of tracking provided by the Publisher and the time until considered not alive set by the Subscriber must be greater than the time set by the Publisher.
+
+Service Owners and Clients will each specify two liveliness policies, one for the liveliness policy pertaining to the Owner and one pertaining to the Client.
+In order for a Client to connect to an Owner to make a request the Client_Liveliness level requested by the Owner must be greater than the level provided by the Client and the Owner_Liveliness requested by the Client must be greater than the level provided by the Owner.
+
+### Lifespan
+
+The lifespan policy establishes a contract for how long a message remains valid.
+The lifespan QoS policy only applies to Topics.
+For Topic Subscribers it establishes the length of time a message is considered valid, after which time it will not be received.
+For Topic Publishers it establishes the length of time a message is considered valid, after which time it will be removed from the Topic history and no longer sent to Subscribers.
+
+- LIFESPAN_DEFAULT - Use the ROS specified default for lifespan (which is LIFESPAN_NONE).
+- LIFESPAN_NONE - Messages do not have a time at which they expire.
+- LIFESPAN_ENABLED - Messages will have a lifespan enforced.
+
+In the future lifespan support could be added for Services to act as a means of timing out requests if they take over a set amount of time.
+At present, that would require additional support for Service Clients to be able to handle timeouts and cleanup requests that is out of scope for this design.
```

Comment:
From what I was seeing in the existing client code, the timeouts on the client side are handled in rclcpp and nothing in the rcl layer or below tracked the time. It's fine to ignore the result when it arrives late, but what happens when it doesn't arrive at all? If for some reason my service is dying or dropping requests am I leaking memory in the rcl/rmw layers by continuing to wait for responses that will never come in? 

Also, as @gbiggs points out, you need a way to notify the application when a timeout occurs. If a request is dropped by either the rmw layer on the client side or by the server the application making the request should be informed of that so that they don't get blocked spinning on a wait to a response that will never come.

Reply:
@nburek: I am not sure if trying to inform application on data removal due to lifespan as sensible for the common use cases of this policy such as:

* Data removal from a transient data writer.
* Data removal from a keep all history data reader.

Removing data and then informing the application of a potential problem created by the data removal makes not really sense to me. A more useful approach would be the latency budget which informs the user that data is older than it should be without removing the data or a way to query the number of samples currently hold in the history per instance.

Reply:
> I am not sure if trying to inform application on data removal due to lifespan as sensible for the common use cases of this policy such as:

@deeplearningrobotics For just Publishers and Subscribers I would agree with you that the application doesn't need to be informed of transient removal of data from the history. However, if you're defining the lifespan for a Service request then the violation of that policy means you never expect to receive a response for the request and you'll need to be informed in both the ROS library to clean up state for the request as well as in the application so that they don't spin forever on waiting for a response that will never come.